aip/auth/4120.md

```diff
@@ -0,0 +1,96 @@
+---
+id: 4120
+state: draft
+created: 2023-12-07
+---
+
+# Universe Domain requirements in Authentication Libraries
+
+This AIP outlines the requirements for Cloud SDK authentication libraries to support Universe Domains, in order to authenticate with APIs outside the GDU (Google Default Universe).
+
+## Guidance
+
+Universe Domain support in the auth library has the following requirements:
+
+ - Credentials implement a `universe_domain` getter
+ - Credentials accept a user-supplied `universe_domain`.
+ - Service Account and External Account credentials read the universe domain from the credentials JSON file.
+ - GCE credentials retrieve the universe domain from the metadata server.
+
+### Naming
+
+We aim to have a consistent user experience across a broad range of clients. One of the invariants is the `universe_domain` name that should be used to represent the universe domain in configuration, user-readable properties and methods. Languages that use different casing conventions should adapt the name to be idiomatic but keep the two specific words in the same order, so `universe-domain`, `universeDomain` and `SetUniverseDomain` are all acceptable, but e.g. `DomainOfUniverse` or `Endpoint` is not.
+
+### Credentials implement a `universe_domain` getter (method, property)
+
+> This requirement applies to **all** flows.
+
+Every credentials class/object **must** expose the `universe_domain` getter (property, method, etc). The auth library end-users, including the client libraries will use this property.
```

Comment:
Can you add some clues about how it should be used?

For example: it is used to verify the credential is the correct one for the universe domain specified by the customer via a client configuration parameter and prevent sending credential information to the wrong endpoints.

Reply:
Sure thing. I'll add something like this:

```
The client libraries will use this property in order to construct the URIs for requests the appropriate universe domain, and to validate that the configured universe domain matches the credentials. 
```